lib/vtls/rustls.c

```diff
@@ -0,0 +1,389 @@
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+#include "curl_setup.h"
+
+#ifdef USE_RUSTLS
+
+#include "curl_printf.h"
+
+#include <errno.h>
+#include <crustls.h>
+
+#include "urldata.h"
+#include "sendf.h"
+#include "vtls.h"
+
+struct ssl_backend_data
+{
+  struct rustls_client_session *session;
+  bool data_pending;
+};
+
+static const struct rustls_client_config *client_config = NULL;
+
+static int
+Curl_rustls_init(void)
+{
+  client_config = rustls_client_config_new();
+  return 1;
+}
+
+static bool
+Curl_rustls_data_pending(const struct connectdata *conn, int sockindex)
+{
+  struct Curl_easy *data = conn->data;
+  const struct ssl_connect_data *connssl = &conn->ssl[sockindex];
+  struct ssl_backend_data *backend = connssl->backend;
+  infof(data, "rustls_data_pending %d\n", backend->data_pending);
+  return backend->data_pending;
+}
+
+static CURLcode
+Curl_rustls_connect(struct connectdata *conn,
+                    int sockindex UNUSED_PARAM)
+{
+  struct Curl_easy *data = conn->data;
+  infof(data, "rustls_connect: unimplemented\n");
+  return CURLE_COULDNT_CONNECT;
+}
+
+/*
+ * On each run:
+ *  - Read a chunk of bytes from the socket into rustls' TLS input buffer.
+ *  - Tell rustls to process any new packets.
+ *  - Read out as many plaintext bytes from rustls as possible, until hitting
+ *    error, EOF, or EAGAIN/EWOULDBLOCK, or plainbuf/plainlen is filled up.
+ *
+ * It's okay to call this function with plainbuf == NULL and plainlen == 0.
+ * In that case, it will copy bytes from the socket into rustls' TLS input
+ * buffer, and process packets, but won't consume bytes from rustls' plaintext
+ * output buffer.
+ */
+static ssize_t
+rustls_recv(struct connectdata *conn, int sockindex, char *plainbuf,
+            size_t plainlen, CURLcode *err)
+{
+  struct Curl_easy *data = conn->data;
+  struct ssl_connect_data *const connssl = &conn->ssl[sockindex];
+  struct ssl_backend_data *const backend = connssl->backend;
+  struct rustls_client_session *const session = backend->session;
+  curl_socket_t sockfd = conn->sock[sockindex];
+  /* Per https://www.bearssl.org/api1.html, max TLS record size plus max
+     per-record overhead. */
+  uint8_t tlsbuf[16384 + 325];
+  ssize_t n = 0;
+  ssize_t tls_bytes_read = 0;
+  ssize_t tls_bytes_processed = 0;
+  ssize_t plain_bytes_copied = 0;
+  int rustls_result = 0;
+
+  memset(tlsbuf, 0, sizeof(tlsbuf));
```

Comment:
Isn't it a bit excessive to clear the entire buffer like this every time?

Reply:
Fair, I've removed it. I did profile curl over a big transfer, and this did not seem to be a significant use of CPU time.

I was looking at the docs for `read` and `recv`, and while implicitly they only write to the provided buf, they don't seem to explicitly guarantee that they don't read from the provided buf, which would be unsafe if the buf was uninitialized. I'm thinking of these related Rust issues:

https://github.com/rust-lang/rust/pull/42002
https://github.com/rust-lang/rust/issues/42788